File: cpp/tennis7.cc
#include <string>

std::string tennis_score(int player1Score, int player2Score)
{
    std::string result = "";
    std::string player1Name = "player1";
    std::string player2Name = "player2";

    if (player1Score == player2Score)
    {
        // tie score
        switch (player1Score)
        {
            case 0: result += "Love-All"; break;
            case 1: result += "Fifteen-All"; break;
            case 2: result += "Thirty-All"; break;
            default: result += "Deuce"; break;
        }
    }
    else if (player1Score >= 4 || player2Score >= 4)
    {
        // end-game score
        if (player1Score - player2Score == 1)
        {
            result += "Advantage " + player1Name;
        }
        else if (player1Score - player2Score == -1)
        {
            result += "Advantage " + player2Name;
        }
        else if (player1Score - player2Score >= 2)
        {
            result += "Win for " + player1Name;
        }
        else
        {
            result += "Win for " + player2Name;
        }
    } else {
        // regular score
        switch (player1Score) {
            case 0:
                result += "Love";
            break;
            case 1:
                result += "Fifteen";
            break;
            case 2:
                result += "Thirty";
            break;
            default:
                result += "Forty";
            break;
        }
        result += "-";
        switch (player2Score) {
            case 0:
                result += "Love";
            break;
            case 1:
                result += "Fifteen";
            break;
            case 2:
                result += "Thirty";
            break;
            default:
                result += "Forty";
            break;
        }
    }

    return result;
}

File: python/tennis7.py
class TennisGame7:
    def __init__(self, player1_name, player2_name):
        self.player1_name = player1_name
        self.player2_name = player2_name
        self.player1_score = 0
        self.player2_score = 0

    def won_point(self, player_name):
        if player_name == "player1":
            self.player1_score += 1
        else:
            self.player2_score += 1

    def score(self):
        result = "Current score: "
        if self.player1_score == self.player2_score:
            # tie score
            result: str
            match self.player1_score:
                case 0:
                    result += "Love-All"
                case 1:
                    result += "Fifteen-All"
                case 2:
                    result += "Thirty-All"
                case _:
                    result += "Deuce"

        elif self.player1_score >= 4 or self.player2_score >= 4:
            # end-game score
            if self.player1_score - self.player2_score == 1:
                result += "Advantage " + self.player1_name
            elif self.player1_score - self.player2_score == -1:
                result += "Advantage " + self.player2_name
            elif self.player1_score - self.player2_score >= 2:
                result += "Win for " + self.player1_name
            else:
                result += "Win for " + self.player2_name

        else:
            # regular score
            match self.player1_score:
                case 0:
                    result += "Love"
                case 1:
                    result += "Fifteen"
                case 2:
                    result += "Thirty"
                case _:
                    result += "Forty"

            result += "-"

            match self.player2_score:
                case 0:
                    result += "Love"
                case 1:
                    result += "Fifteen"
                case 2:
                    result += "Thirty"
                case _:
                    result += "Forty"

        return result + ", enjoy your game!"
